certora/specs/ERC20.spec

```diff
@@ -1,61 +1,46 @@
 // This is spec is taken from the Open Zeppelin repositories at https://github.com/OpenZeppelin/openzeppelin-contracts/blob/448efeea6640bbbc09373f03fbc9c88e280147ba/certora/specs/ERC20.spec, and patched to support the DelegationToken.
 // Note: the scope of this specification encompases only the implementation contract and not the interactions with the implementation contract and the proxy.
 
+import "ERC20Invariants.spec";
+
 definition nonpayable(env e) returns bool = e.msg.value == 0;
 definition nonzerosender(env e) returns bool = e.msg.sender != 0;
 
 methods {
     function name()                                external returns (string)  envfree;
     function symbol()                              external returns (string)  envfree;
     function decimals()                            external returns (uint8)   envfree;
-    function totalSupply()                         external returns (uint256) envfree;
-    function balanceOf(address)                    external returns (uint256) envfree;
     function allowance(address,address)            external returns (uint256) envfree;
     function owner()                               external returns address   envfree;
-    function delegatedVotingPower(address)         external returns uint256   envfree;
-    function delegatee(address)                    external returns address   envfree;
     function nonces(address)                       external returns (uint256) envfree;
     function DOMAIN_SEPARATOR()                    external returns (bytes32) envfree;
-    function upgradeToAndCall(address, bytes)      external => NONDET DELETE;
 }
 
 /*
 ┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
-│ Ghost & hooks: sum of all balances                                                                                  │
+│ Invariant: totalSupply is the sum of all balances                                                                   │
 └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
 */
-ghost mathint sumOfBalances {
-    init_state axiom sumOfBalances == 0;
-}
 
-// Slot is ERC20Storage._balances slot
-hook Sload uint256 balance (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address addr] {
-   require sumOfBalances >= to_mathint(balance);
-}
+use invariant sumOfBalancesStartsAtZero;
```

Comment:
I'm not really comfortable with this pattern of `use invariant`. I understand it is functionally the same, but it introduces the possibility to miss one invariant. Plus it adds lines of code that are not really necessary IMO. Can we just prove `ERC20Invariant.spec` instead, and not import invariants this way ?

Reply:
The rationale at the time of development was that, the definition of invariants needs to be shared with other spec (i. e. Delegation.spec) in particular the definition of ghost variables and hooks. At the same time, there was still the concern to keep some proximity with the original OZ spec. The idea was to keep everything that brings too invasive changes (w.r.t the original source file), I didn't think of the prelude of shared definitions as being "checkable" but merely a library of definitions. We can definitely add configuration files to check those.

Reply:
As a compromise, I added the config files in https://github.com/morpho-org/morpho-token-upgradeable/pull/102/commits/b6dbaec16e4fa903589e1aed9a4519ea9fe6ef21 and left only the invariant of interest in the ERC20 spec (note: it's now being checked twice).